### PR TITLE
Implement copy and regenerate action of chat panel.

### PR DESCRIPTION
**Pull request by boyan-tonchev**
## What
Added functionality for copying and regenerating chat content:
 - The copy action allows users to copy chat answer to the clipboard;
 - The regenerate action re-generates the answer of a question.

## Why
To fulfill the requirements for user interactions with chat content.

## How
The copy and regenerate actions are implemented in the main TTYG controller. Actions are triggered from the chat panel. Communication between the chat panel component and the controller is managed through the TTYG Context Service.

### CODE REVIEW

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 1]
```diff
@@ -64,10 +64,21 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
                 if (!$scope.chatItem.chatId) {
                     createNewChat();
                 } else {
-                    askQuestion();
+                    askQuestion($scope.chatItem);
                 }
             };
 
+            $scope.regenerate = (chatItem) => {
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:71]: write jsdoc for the public methods and fields at least!

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 2]
```diff
@@ -64,10 +64,21 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
...
+                const question = getEmptyChatItem();
+                question.question.message = chatItem.question.message;
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:73]: I don't like this. You are revealing the model too much.

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 3]
```diff
@@ -64,10 +64,21 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
                 if (!$scope.chatItem.chatId) {
                     createNewChat();
                 } else {
-                    askQuestion();
+                    askQuestion($scope.chatItem);
                 }
             };
 
+            $scope.regenerate = (chatItem) => {
+                const question = getEmptyChatItem();
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:72]: This is a ChatItemModel and not a question

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 4]
```diff
@@ -64,10 +64,21 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
...
+                question.question.message = chatItem.question.message;
+                const regenerateChatItem = new ChatItemModel(chatItem.chatId, chatItem.agentId, question.question);
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:74]: What's the point to create a new ChatItemModel just to get the question from it?!

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 5]
```diff
@@ -64,10 +64,21 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
...
+                askQuestion(regenerateChatItem);
+            };
+
+            $scope.copy = (chatItem) => {
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:78]: `copy` what?

File: src/js/angular/ttyg/directives/chat-panel.directive.js [hunk 6]
```diff
@@ -103,6 +114,14 @@ function ChatPanelComponent(toastr, $translate, TTYGContextService, TTYGService)
                 scrollToBottom();
             };
 
+            const onCopied = () => {
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/directives/chat-panel.directive.js:117]: onCopied what?

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 1 above)
Comment (svilenvelikov) [reply src/js/angular/ttyg/directives/chat-panel.directive.js:71]: Also, can you be more specific in the methods naming.

File: src/js/angular/ttyg/templates/chat-panel.html [hunk 7]
```diff
@@ -34,7 +42,7 @@
         <input class="form-control question-input" ng-model="chatItem.question.message" type="text">
         <button class="btn btn-primary ask-button" ng-disabled="!chatItem.question.message" ng-click="ask()">
             <i class="fa-regular fa-arrow-turn-left-up"></i>
-            {{ 'ttyg.chat_panel.btn.aks' | translate }}
+            {{ 'ttyg.chat_panel.btn.aks.label' | translate }}
```
Comment (svilenvelikov) [inline src/js/angular/ttyg/templates/chat-panel.html:45]: typo

**Review by svilenvelikov — CHANGES_REQUESTED** [review]

File: src/js/angular/ttyg/templates/chat-panel.html (hunk 7 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/templates/chat-panel.html]: Fixed

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 1 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js]: The name has been changed, and Javadoc has been added.

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 2 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js]: Added methods to the model that fetch and update its message.

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 3 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js]: Renamed.

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 4 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js]: Oops, this is an old implementation from before I asked which agentId to use: the chat item's agent or the selected agent. I forgot to update this part of code.

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 5 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js]: Renamed

File: src/js/angular/ttyg/directives/chat-panel.directive.js (hunk 6 above)
Comment (boyan-tonchev) [reply src/js/angular/ttyg/directives/chat-panel.directive.js:126]: Renamed.

**Review by svilenvelikov — APPROVED** [review]

**Comment by boyan-tonchev** [comment]
Jenkins retry build

**Comment by boyan-tonchev** [comment]
Jenkins retry build

**Review by svilenvelikov — APPROVED** [review]

**Comment by boyan-tonchev** [comment]
Jenkins retry build

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**